### DISCUSSION TITLE
Using Browsers with HasRelated behavior

### DISCUSSION THREAD

**Discussion by nikhiltri** [area17/twill#940] [General]
As part of upgrading to the latest version of Twill we're working on cleaning up our repositories to use [Twill's new generalized browser functionality](https://github.com/area17/twill/blob/2.x/src/Repositories/Behaviors/HandleBrowsers.php#L10-L29). One thing we're finding is that this functionality only works when the model uses a proper relation, e.g., the `updateBrowser` method looks for a model method that uses an Eloquent relationship [that it can run a `sync()` on](https://github.com/area17/twill/blob/2.x/src/Repositories/Behaviors/HandleBrowsers.php#L77).

For some of our model relationships, we use the [HasRelated](https://github.com/area17/twill/blob/2.x/src/Models/Behaviors/HasRelated.php) behavior so we don't have to build out DB tables for something minor. But I'm not sure this new browser functionality can work with relationships that use HasRelated. 

I'm not sure what a good solution might be. Perhaps in the array passed in the definition in the `$browser` array, we can say whether the regular browser methods should be called, or the related methods? E.g., `updateBrowser` vs. `updateRelatedBrowser`, and `getFormFieldsForBrowser` vs. `getFormFieldsForRelatedBrowser`.

Is this something you've already accounted for in some way? Or is this something new?

**Comment by ifox** [comment]
Hi @nikhiltri! 

This is a great suggestion, both because it is indeed not currently implemented and because the approach you propose makes sense. We can add a new property in the `browsers` array items to elect to use the `HasRelated` specific functions.

**Comment by pboivin** [comment]
Hi @nikhiltri,

Curious to know if you have implemented this on your end. Would it be OK if I submitted a PR for this?

**Reply by nikhiltri** [reply]
Hey @pboi20,

We created a new behavior in our codebase that replicates some of the functionality of `HandleBrowser`. We didn't actually change anything is `HandleBrowser` but do reference methods defined there. I'm not sure if this is ultimately the best route to take:

```
<?php

namespace App\Repositories\Behaviors;

/**
 * Mimic Twill's generalized behavior for browsers,
 * but when the relation uses HasRelated instead of being
 * a proper model relation.
 *
 * @see A17\Twill\Repositories\Behaviors\HandleBrowsers
 * @see https://github.com/area17/twill/discussions/940
 */
trait HandleRelatedBrowsers
{
    protected $relatedBrowsers = [];

    protected function getRelatedBrowsers()
    {
        return collect($this->relatedBrowsers)->map(function ($browser, $key) {
            $browserName = is_string($browser) ? $browser : $key;
            $moduleName = !empty($browser['moduleName']) ? $browser['moduleName'] : $this->inferModuleNameFromBrowserName($browserName);

            return [
                'relation' => !empty($browser['relation']) ? $browser['relation'] : $this->inferRelationFromBrowserName($browserName),
                'model' => !empty($browser['model']) ? $browser['model'] : $this->inferModelFromModuleName($moduleName),
                'browserName' => $browserName,
            ];
        })->values();
    }

    public function afterSaveHandleRelatedBrowsers($object, $fields)
    {
        foreach ($this->getRelatedBrowsers() as $browser) {
            $this->updateRelatedBrowser($object, $fields, $browser['browserName']);
        }
    }

    public function getFormFieldsHandleRelatedBrowsers($object, $fields)
    {
        foreach ($this->getRelatedBrowsers() as $browser) {
            $fields['browsers'][$browser['browserName']] = $this->getFormFieldsForRelatedBrowser($object, $browser['relation']);
        }

        return $fields;
    }
}
```

This works because it plugs into ModuleRepository's [afterSave()](https://github.com/area17/twill/blob/29f7364c2fd7aa0f6cfccb05bf105087fea793de/src/Repositories/ModuleRepository.php#L599-L604) and [getFormField()](https://github.com/area17/twill/blob/29f7364c2fd7aa0f6cfccb05bf105087fea793de/src/Repositories/ModuleRepository.php#L599-L604) functions.

**Reply by pboivin** [reply]
@nikhiltri, this is great! I find the idea of a standalone behavior cleaner than [what I had in mind](https://github.com/pboi20/twill/commit/9bea00812e83a0cb22c97f4ce1764b4f62027c67). Thanks for sharing :)

**Reply by ifox** [reply]
That's great, guys! Will happily merge a PR adding this functionality.

**Reply by nikhiltri** [reply]
Here you go! https://github.com/area17/twill/pull/967
